### PR TITLE
avm2: Introduce `TextFormatObject`

### PR DESCRIPTION
**Pull request by relrelb**
Similar to #5569, just for AVM2.

### CODE REVIEW

File: core/src/avm2/object/textformat_object.rs [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        let proto = activation.avm2().prototypes().textformat;
+        let base = ScriptObjectData::base_new(Some(proto), Some(class));
+
+        let mut this: Object<'gc> = Self(GcCell::allocate(
+            activation.context.gc_context,
+            TextFormatObjectData { base, text_format },
+        ))
+        .into();
+        this.install_instance_traits(activation, class)?;
+
```
Comment (relrelb) [inline core/src/avm2/object/textformat_object.rs:58]: Note: I've seen many functions (e.g. `RegExpObject::from_regexp`) do the following:
```rs
class.call_native_init(Some(this), &[], activation)?;
```
But I think this is wrong in this case, as it applies the `TextFormat` constructor on the newly-created `this` with no parameters, which deletes its fields.
Maybe I'm missing something, but the existing `call_native_init` calls seem redundant, and even wrong, to me.

Comment (kmeisthax) [reply core/src/avm2/object/textformat_object.rs:58]: The pattern I adopted with `StageObject` and a few other object types was that `this` holds nullable data and the initializer only changes anything if that data is `None`. `RegExpObject` does not have a nullable `regexp` field, so this pattern doesn't work. Indeed, you are correct that calling `from_regexp` will not actually use the given `RegExp` correctly. However, this function is apparently unused, which is why this doesn't get flagged by any tests.

Generally speaking associated constructors (such as `from_regexp`) do need to call the class initializer, in one of two cases:

1. The class the object is being allocated as is user-specified.
2. One or more of the object's superclasses (most notably, `EventDispatcher`) does something not accounted for by the associated constructor.

AFAIK neither condition applies for `TextFormatObject` or `RegExpObject`. I added calls to `call_native_init` everywhere for the sake of consistency (e.g. every object allocation also runs class initializers); however it probably would be permissible to drop this for specific classes as long as we document that initializers are being skipped.

**Review by kmeisthax — APPROVED** [review]

**Comment by adrian17** [comment]
Sorry for a silly late question (copied from Discord), but:

> TextFormat objects differ from regular objects in that
TextField.setTextFormat and TextField.setNewTextFormat accept
only the former, and ignore the latter.

Is this the only motivation for these two PRs? Because AS-level typechecking sounds like a relatively small justification for such a big added Rust-internal object type, I think...? (that said, I don't know any other way yet).
(Also these didn't add any tests for what was supposed to change :/ )

I also didn't see this question from the original PR answered:

> I'm going to assume the extra type checks in AS3 make an AVM2 version of this object type unnecessary? Or should we be doing that, too?

**Comment by relrelb** [comment]
> Sorry for a silly late question (copied from Discord), but:
> 
> > TextFormat objects differ from regular objects in that
> > TextField.setTextFormat and TextField.setNewTextFormat accept
> > only the former, and ignore the latter.
> 
> Is this the only motivation for these two PRs? Because AS-level typechecking sounds like a relatively small justification for such a big added Rust-internal object type, I think...? (that said, I don't know a good way yet... maybe keep it an enum variant, but make it a thin layer that still just passes-through the values to the underlying ScriptObject?). (Also these didn't add any tests for what was supposed to change :/ )
> 
> I also didn't see this question from the original PR answered:
> 
> > I'm going to assume the extra type checks in AS3 make an AVM2 version of this object type unnecessary? Or should we be doing that, too?

Sorry for the unresponsiveness, the main motiviation for this change was indeed the AS-level typecheckkng and introduction of a more accurate behavior. A secondary motivation is organizing AVM1 and AVM2 code into their respective directories, avoiding AVM-specific code in `core` as much as possible. This brings us one step closer to self-hosted AVM components, that has many advantages (see https://github.com/ruffle-rs/ruffle/pull/5288#pullrequestreview-752385590).
As for tests, you're totally right. I'm going to add those in the next few days.
And actually this PR answers kmeisthax's question, probably I needed to answer it right there.

**Pull request merged**